doc/cookbook/current-route/CurrentRoute.lhs

```diff
@@ -0,0 +1,391 @@
+        BS8.putStrLn p
+        return (succ i)
+    }
+
+app1 :: Application
+app1 = genericServe routes1
+```
+
+We define a `Routes1` data type, implement `Routes1 AsServer` value and turn it
+into the `wai`'s `Application`. However, to compile this example, we need an
+additional instance.  We use an *internal* `passToServer` combinator in the
+implementation of `route`.
+
+```haskell
+instance HasServer api ctx => HasServer (Wai.Request :> api) ctx where
+    type ServerT (Wai.Request :> api) m = Wai.Request -> ServerT api m
+    
+    hoistServerWithContext _ pc nt s =
+        hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+    route _ ctx d = route (Proxy :: Proxy api) ctx $
+        passToServer d id
+```
+
+This solution is good quick fix, but there are arguably better ways.
+
+### Specific combinator
+
+We may notice that both our handlers use `Wai.rawPathInto req` call.
+That should alert us. Specific combinator is more elegant.
+An ability to create new combinators outside the core framework,
+is one of design principles of `servant`.
+
+```haskell
+data RawPathInfo
+
+instance HasServer api ctx => HasServer (RawPathInfo :> api) ctx where
+    type ServerT (RawPathInfo :> api) m = BS8.ByteString -> ServerT api m
+    
+    hoistServerWithContext _ pc nt s =
+        hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+    route _ ctx d = route (Proxy :: Proxy api) ctx $
+        passToServer d Wai.rawPathInfo
+```
+
+Using new `RawPathInfo` combinator, we can re-implement our application:
+
+```haskell
+data Routes2 route = Routes2
+    { route21 :: route :- RawPathInfo :> "route1" :> Get '[JSON] Int
+    , route22 :: route :- RawPathInfo :> "route2" :> Capture "id" Int :> Get '[JSON] Int
+    }
+  deriving (Generic)
+
+routes2 :: Routes2 AsServer
+routes2 = Routes2
+    { route21 = \p -> liftIO $ do
+        BS8.putStrLn p
+        return (BS8.length p)
+    , route22 = \p i -> liftIO $ do
+        BS8.putStrLn p
+        return (succ i)
+    }
+
+app2 :: Application
+app2 = genericServe routes2
+```
+
+This version is slightly more declarative, and handlers are more restrictive.
+We moved the `rawPathInfo` selector from handlers to combinator implementation,
+removed repetition.
+
+### Using `Vault`
+
+The `vault` value in `wai` `Request` is not well known or used.
+But in this scenario it can be useful.
+Vault is explained in [Using WAI's vault for fun and profit](https://www.yesodweb.com/blog/2015/10/using-wais-vault) blog post.
+It fills a "dynamic" gap of strongly typed `Request`: we can attach arbitrary data to the request,
+as is common in web frameworks in a dynamically typed languages.
+As `servant-server` is based on `wai`, using vault is the third answer
+to the first part of the question.
+
+We (unsafely) create a key to the vault:
+
+```haskell
+rpiKey :: V.Key BS8.ByteString
+rpiKey = unsafePerformIO V.newKey
+```
+
+Then we create a *middleware* which will put `rawPathInfo` into the `vault`.
```

Comment:
Would be great to have the "middleware" bit link to some place that documents them (well, ideally -- does the yesod book have a section about them?).